### PR TITLE
[networks] Fix HTTPS overcount

### PR DESCRIPTION
**Pull request by p-lambert**
### What does this PR do?

Fix an issue that was resulting in an overcount of HTTPS requests when tracing multiple instances of apache (`httpd`) servers in the same node.

### Additional Notes

During the investigation of this problem we identified 2 distinct issues:

#### OpenSSL/GNUTLS library call sequence

One issue has to do with the way `httpd` integrates with OpenSSL/GNUTLS. For a reason that still eludes me (I haven't gone over the actual source code of the apache module `mod_ssl.so`), each incoming HTTPS request (say, `GET /foo HTTP/1.1`) results in a rather unusual sequence of `SSL_{read,write}` calls:

1. `SSL_read (GET /foo HTTP/1.1)`
2. `SSL_write (HTTP/1.1 200 OK)`
3. `SSL_read (GET /foo HTTP/1.1)`
4. `SSL_read (GET /foo HTTP/1.1)`

This breaks the HTTP eBPF "state-machine" code because when function call `3` is called the program thinks that a new request is being made for a keep-alive socket. It will then flush data corresponding to 1&2 and will start tracking a new HTTP transaction (`http_begin_request`). This is where the bug comes in: we would only override the request part (`request_fragment` and `request_method`), but we would leave `http_response_status` among other response fields with their previous values from `2`.
Finally, when `4` is called, the eBPF program would see think that again a new request was starting, and would flush _again_ a transaction to userspace, which resulted in the overcount.

#### Multiple instances of the same container

Prior to this change we were effectively attaching the same uprobes multiple times when running multiple instances of the same container in the same host. This was happening because each container instance would have its own (overlay) filesytem, with its own unique shared library paths, but they actually share the same underlying inodes. In this change we're refactoring `http/shared_libraries.go` to take inodes into account.

#### Step 1: spin up 2 apache instances

* `docker run --rm -d -p 443:443 pedrolambert/apache-mod_ssl | xargs docker inspect | jq '.[0].NetworkSettings.IPAddress'`
* `docker run --rm -d pedrolambert/apache-mod_ssl`

This docker image basically enables `mod_ssl` for apache and configures a self-signed certificate.
Only one container needs to bind to port 443; The other one is merely to trigger another shared library load.

#### Step 2: Issue HTTPS request and verify request count reported by `system-probe`

* `curl -I -k https://127.0.0.1:443`
* `sudo curl -s --unix /opt/datadog-agent/run/sysprobe.sock http://unix/network_tracer/debug/http_monitoring | jq '.[] | select(.Server.IP!="127.0.0.1" and .Server.Port==443)'`
Note: double check the `Count` field and make sure it's `1`

#### Step 3: Cleanup env

* `docker ps | grep apache | awk '{print $1}' | xargs docker stop`
* `docker rmi pedrolambert/apache-mod_ssl`

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by brycekahle — CHANGES_REQUESTED** [review]
I'd like to see two tests:
1. One asserting this fix for the overcount
2. One asserting the library fix.

In addition, it was clear what the problem was, but it was not clear how we are fixing it.

**Comment by p-lambert** [comment]
> I'd like to see two tests:
> 
> 1. One asserting this fix for the overcount
> 2. One asserting the library fix.
> 
> In addition, it was clear what the problem was, but it was not clear how we are fixing it.

Added tests for `2.` and QA steps for `1.` as reproducing it programmatically it too hard

File: pkg/network/http/shared_libraries_test.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+)
+
+func TestSharedLibraryDetection(t *testing.T) {
+	skipTestIfKernelNotSupported(t)
+	if !httpsSupported() {
+		t.Skip("shared library detection not supported. skipping.")
+	}
+
+	perfHandler, doneFn := initEBPFProgram(t)
+	fpath := filepath.Join(os.TempDir(), "foo.so")
```
Comment (brycekahle) [inline pkg/network/http/shared_libraries_test.go:40]: nit: use `t.TempDir()` and you don't need to worry about removing the file afterwards.

**Review by brycekahle — APPROVED** [review]

File: pkg/network/http/shared_libraries_test.go (hunk 1 above)
Comment (p-lambert) [reply pkg/network/http/shared_libraries_test.go:35]: nice!

**Comment by DanielLavie** [comment]
Do you have the pedrolambert/apache-mod_ssl image for arm as well?

**Pull request merged**